### ISSUE TITLE
Pylance will ignore anything marked as Never

### ISSUE THREAD

**Issue by llacroix** [microsoft/pyright#6512]
## Environment data

-   OS and version: Linux
-   Python version (& distribution if applicable, e.g. Anaconda): 3.9

## Code Snippet

```python
class Test(object):
    __slots__ = ['env']

    def __init__(self):
        pass

    def search(self, val:int) -> int:
        return val + 1

def func(val:int):
    x = Test()
    obj: Test = x.env['abc']
    print(obj.search(val + vals))
```

## Repro Steps

The variable `val` will be marked as unused and the vals undefined variable isn't raised as an error.

## Expected behavior

At the very least, pylance should comply with the code that's written. So obj should be of the type Test because it's hinted as such. Then the val should be seen as used because it's used in the search method. The vals should raise an error because it's not there anywhere.

## Actual behavior

The env slot has the Never type and anything touching this gets the Never type regardless of the hints assigned. For example, here the obj is also of the Never type... then any method call on the obj is of the never type... which means that val is never used and pylance doesn't even bother checking that vals exists or not. It doesn't make sense for pylance to ignore code because of the Never type because in Python, it will either raise an exception or return None. It's not like Python is smart enough to detect and remove "unreachable code" as far as I know. 

But what's super strange is that if you use `getattr(x, 'env')` it will return the `Any` type instead of `Never`.

**Comment by erictraut** [comment]
Thanks for reporting the issue. I agree there's a bug here, but your analysis isn't entirely correct.

The problem here is that `env` is implicitly declared (via the `__slots__` assignment), but its type is not explicitly declared, nor is it ever assigned a value, so its type cannot be inferred. That means pyright infers its type as "Unbound". You can see this if you hover over `env`.

The bug is that an unbound value, when used on the LHS (left-hand side) of an index expression should generate a type of `Unknown` (an implicit `Any`) rather than `Never`.

This change will be included in the next release of pyright and a future release of pylance.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.337, which I just published. It will also be included in a future release of pylance.

**Comment by llacroix** [comment]
> Thanks for reporting the issue. I agree there's a bug here, but your analysis isn't entirely correct.
> 
> The problem here is that `env` is implicitly declared (via the `__slots__` assignment), but its type is not explicitly declared, nor is it ever assigned a value, so its type cannot be inferred. That means pyright infers its type as "Unbound". You can see this if you hover over `env`.
> 
> The bug is that an unbound value, when used on the LHS (left-hand side) of an index expression should generate a type of `Unknown` (an implicit `Any`) rather than `Never`.
> 
> This change will be included in the next release of pyright and a future release of pylance.

Yeah the slot is the only way I could reproduce this. But thanks, that will surely help.
